File: maplibre_gl_example/android/app/src/main/java/org/maplibre/example/terrain/LercDecoderPlugin.java
package org.maplibre.example.terrain;

import androidx.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import java.util.HashMap;
import java.util.Map;

/**
 * FlutterPlugin that bridges Flutter and native LERC decoder.
 */
public class LercDecoderPlugin implements FlutterPlugin, MethodCallHandler {
    private MethodChannel channel;
    private LercNativeLoader lercNativeLoader;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPlugin.FlutterPluginBinding binding) {
        channel = new MethodChannel(binding.getBinaryMessenger(), "org.maplibre.example/lerc_decoder");
        channel.setMethodCallHandler(this);
        
        try {
            lercNativeLoader = LercNativeLoader.getInstance();
        } catch (RuntimeException e) {
            System.err.println("Failed to initialize LERC native library: " + e.getMessage());
        }
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPlugin.FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
        channel = null;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (lercNativeLoader == null) {
            result.error("LERC_NOT_INITIALIZED", "LERC native library failed to initialize", null);
            return;
        }

        switch (call.method) {
            case "getLercInfo":
                byte[] infoBuffer = call.argument("buffer");
                if (infoBuffer == null) {
                    result.error("INVALID_ARGUMENT", "Buffer cannot be null", null);
                    return;
                }
                
                LercInfo info = lercNativeLoader.getLercInfo(infoBuffer);
                if (info == null) {
                    result.error("DECODE_ERROR", "Failed to get LERC info", null);
                    return;
                }
                
                Map<String, Object> infoMap = new HashMap<>();
                infoMap.put("width", info.width);
                infoMap.put("height", info.height);
                infoMap.put("numBands", info.numBands);
                infoMap.put("numValidPixels", info.numValidPixels);
                infoMap.put("minValue", info.minValue);
                infoMap.put("maxValue", info.maxValue);
                infoMap.put("noDataValue", info.noDataValue);
                
                result.success(infoMap);
                break;
                
            case "decodeLerc":
                byte[] decodeBuffer = call.argument("buffer");
                Map<String, Object> infoArg = call.argument("info");
                
                if (decodeBuffer == null || infoArg == null) {
                    result.error("INVALID_ARGUMENT", "Buffer and info cannot be null", null);
                    return;
                }
                
                // Create LercInfo from the passed map
                LercInfo lercInfo = new LercInfo(
                    (int) infoArg.get("width"),
                    (int) infoArg.get("height"),
                    (int) infoArg.get("numBands"),
                    (int) infoArg.get("numValidPixels"),
                    (double) infoArg.get("minValue"),
                    (double) infoArg.get("maxValue"),
                    (double) infoArg.get("noDataValue")
                );
                
                double[] decodedData = lercNativeLoader.decodeLerc(decodeBuffer, lercInfo);
                if (decodedData == null) {
                    result.error("DECODE_ERROR", "Failed to decode LERC data", null);
                    return;
                }
                
                result.success(decodedData);
                break;
                
            default:
                result.notImplemented();
                break;
        }
    }
}

File: maplibre_gl_example/ios/Runner/lerc_wrapper/LercDecoderPlugin.m
#import <Foundation/Foundation.h>
#import <Flutter/Flutter.h>
#import "LercDecoder.h"

@interface LercDecoderPlugin : NSObject<FlutterPlugin>
@end

@implementation LercDecoderPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"org.maplibre.example/lerc_decoder"
                                     binaryMessenger:[registrar messenger]];
    LercDecoderPlugin* instance = [[LercDecoderPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"getLercInfo" isEqualToString:call.method]) {
        FlutterStandardTypedData *buffer = call.arguments[@"buffer"];
        if (buffer == nil) {
            result([FlutterError errorWithCode:@"INVALID_ARGUMENT"
                                       message:@"Buffer cannot be null"
                                       details:nil]);
            return;
        }
        
        LercInfo *info = [[LercDecoder sharedInstance] getInfoFromData:buffer.data];
        if (info == nil) {
            result([FlutterError errorWithCode:@"DECODE_ERROR"
                                       message:@"Failed to get LERC info"
                                       details:nil]);
            return;
        }
        
        NSDictionary *infoMap = @{
            @"width": @(info.width),
            @"height": @(info.height),
            @"numBands": @(info.numBands),
            @"numValidPixels": @(info.numValidPixels),
            @"minValue": @(info.minValue),
            @"maxValue": @(info.maxValue),
            @"noDataValue": @(info.noDataValue)
        };
        
        result(infoMap);
    }
    else if ([@"decodeLerc" isEqualToString:call.method]) {
        FlutterStandardTypedData *buffer = call.arguments[@"buffer"];
        NSDictionary *infoArg = call.arguments[@"info"];
        
        if (buffer == nil || infoArg == nil) {
            result([FlutterError errorWithCode:@"INVALID_ARGUMENT"
                                       message:@"Buffer and info cannot be null"
                                       details:nil]);
            return;
        }
        
        // Create LercInfo from the passed map
        LercInfo *info = [[LercInfo alloc] init];
        [info setValue:infoArg[@"width"] forKey:@"width"];
        [info setValue:infoArg[@"height"] forKey:@"height"];
        [info setValue:infoArg[@"numBands"] forKey:@"numBands"];
        [info setValue:infoArg[@"numValidPixels"] forKey:@"numValidPixels"];
        [info setValue:infoArg[@"minValue"] forKey:@"minValue"];
        [info setValue:infoArg[@"maxValue"] forKey:@"maxValue"];
        [info setValue:infoArg[@"noDataValue"] forKey:@"noDataValue"];
        
        NSArray<NSNumber *> *decodedData = [[LercDecoder sharedInstance] decodeData:buffer.data withInfo:info];
        if (decodedData == nil) {
            result([FlutterError errorWithCode:@"DECODE_ERROR"
                                       message:@"Failed to decode LERC data"
                                       details:nil]);
            return;
        }
        
        result(decodedData);
    }
    else {
        result(FlutterMethodNotImplemented);
    }
}

@end
